File: hofmann-rfc/src/main/java/com/codeheadsystems/rfc/opaque/internal/OpaqueAke.java
package com.codeheadsystems.rfc.opaque.internal;

import com.codeheadsystems.rfc.common.ByteUtils;
import com.codeheadsystems.rfc.ellipticcurve.rfc9380.GroupSpec;
import com.codeheadsystems.rfc.opaque.config.OpaqueCipherSuite;
import com.codeheadsystems.rfc.opaque.config.OpaqueConfig;
import com.codeheadsystems.rfc.opaque.internal.OpaqueEnvelope.RecoverResult;
import com.codeheadsystems.rfc.opaque.model.AuthResult;
import com.codeheadsystems.rfc.opaque.model.ClientAuthState;
import com.codeheadsystems.rfc.opaque.model.CredentialRequest;
import com.codeheadsystems.rfc.opaque.model.CredentialResponse;
import com.codeheadsystems.rfc.opaque.model.KE1;
import com.codeheadsystems.rfc.opaque.model.KE2;
import com.codeheadsystems.rfc.opaque.model.KE3;
import com.codeheadsystems.rfc.opaque.model.RegistrationRecord;
import com.codeheadsystems.rfc.opaque.model.ServerAuthState;
import com.codeheadsystems.rfc.opaque.model.ServerKE2Result;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Arrays;

/**
 * OPAQUE-3DH Authenticated Key Exchange implementation.
 */
public class OpaqueAke {

  private OpaqueAke() {
  }

  /**
   * Constructs the preamble for key derivation.
   * preamble = "OPAQUEv1-" || encode_vector(context) || encode_vector(clientIdentity)
   * || serialize(ke1) || encode_vector(serverIdentity)
   * || serialize(credentialResponse) || serverNonce || serverAkePublicKey
   *
   * @param context            the context
   * @param clientIdentity     the client identity
   * @param ke1                the ke 1
   * @param serverIdentity     the server identity
   * @param credentialResponse the credential response
   * @param serverNonce        the server nonce
   * @param serverAkePublicKey the server ake public key
   * @return the byte [ ]
   */
  public static byte[] buildPreamble(byte[] context, byte[] clientIdentity, KE1 ke1,
                                     byte[] serverIdentity, CredentialResponse credentialResponse,
                                     byte[] serverNonce, byte[] serverAkePublicKey) {
    byte[] prefix = "OPAQUEv1-".getBytes(StandardCharsets.US_ASCII);
    return ByteUtils.concat(
        prefix,
        encodeVector(context),
        encodeVector(clientIdentity),
        ke1.serialize(),
        encodeVector(serverIdentity),
        serializeCredentialResponse(credentialResponse),
        serverNonce,
        serverAkePublicKey
    );
  }

  /**
   * Derives session keys from ikm and preamble using HKDF-Expand-Label.
   */
  private static DerivedKeys deriveKeys(OpaqueConfig config, byte[] ikm, byte[] preamble) {
    OpaqueCipherSuite suite = config.cipherSuite();
    byte[] prk = suite.hkdfExtract(new byte[0], ikm);
    byte[] preambleHash = suite.hash(preamble);

    byte[] handshakeSecret = suite.hkdfExpandLabel(prk,
        "HandshakeSecret".getBytes(StandardCharsets.US_ASCII), preambleHash, config.Nx());
    byte[] sessionKey = suite.hkdfExpandLabel(prk,
        "SessionKey".getBytes(StandardCharsets.US_ASCII), preambleHash, config.Nx());
    Arrays.fill(prk, (byte) 0);

    byte[] km2 = suite.hkdfExpandLabel(handshakeSecret,
        "ServerMAC".getBytes(StandardCharsets.US_ASCII), new byte[0], config.Nm());
    byte[] km3 = suite.hkdfExpandLabel(handshakeSecret,
        "ClientMAC".getBytes(StandardCharsets.US_ASCII), new byte[0], config.Nm());
    Arrays.fill(handshakeSecret, (byte) 0);

    return new DerivedKeys(km2, km3, sessionKey);
  }

  /**
   * Server GenerateKE2.
   *
   * @param config               OPAQUE configuration (provides context and cipher suite)
   * @param serverIdentity       server identity bytes (or serverPublicKey if null)
   * @param serverPrivateKey     server long-term private key
   * @param serverPublicKey      server long-term public key bytes
   * @param record               stored registration record
   * @param credentialIdentifier credential identifier
   * @param oprfSeed             server OPRF seed
   * @param ke1                  client's KE1 message
   * @param clientIdentity       client identity bytes (or clientPublicKey if null)
   * @param maskingNonce         provided masking nonce (null = random)
   * @param serverAkeKeySeed     provided server ephemeral AKE seed (null = random)
   * @return ServerKE2Result containing serverAuthState and ke2
   */
  public static ServerKE2Result generateKE2(OpaqueConfig config, byte[] serverIdentity,
                                            BigInteger serverPrivateKey, byte[] serverPublicKey,
                                            RegistrationRecord record, byte[] credentialIdentifier,
                                            byte[] oprfSeed, KE1 ke1, byte[] clientIdentity,
                                            byte[] maskingNonce, byte[] serverAkeKeySeed) {
    byte[] resolvedSeed = (serverAkeKeySeed != null) ? serverAkeKeySeed
        : config.randomProvider().randomBytes(OpaqueConfig.Nn);
    byte[] serverNonce = config.randomProvider().randomBytes(OpaqueConfig.Nn);
    return generateKE2Deterministic(config, serverIdentity, serverPrivateKey, serverPublicKey,
        record, credentialIdentifier, oprfSeed, ke1, clientIdentity, maskingNonce, resolvedSeed, serverNonce);
  }

  /**
   * Server GenerateKE2 with deterministic server nonce (for testing).
   *
   * @param config               the config
   * @param serverIdentity       the server identity
   * @param serverPrivateKey     the server private key
   * @param serverPublicKey      the server public key
   * @param record               the record
   * @param credentialIdentifier the credential identifier
   * @param oprfSeed             the oprf seed
   * @param ke1                  the ke 1
   * @param clientIdentity       the client identity
   * @param maskingNonce         the masking nonce
   * @param serverAkeKeySeed     the server ake key seed
   * @param serverNonce          the server nonce
   * @return the server ke 2 result
   */
  public static ServerKE2Result generateKE2Deterministic(OpaqueConfig config, byte[] serverIdentity,
                                                         BigInteger serverPrivateKey, byte[] serverPublicKey,
                                                         RegistrationRecord record, byte[] credentialIdentifier,
                                                         byte[] oprfSeed, KE1 ke1, byte[] clientIdentity,
                                                         byte[] maskingNonce, byte[] serverAkeKeySeed,
                                                         byte[] serverNonce) {
    OpaqueCipherSuite suite = config.cipherSuite();

    byte[] sId = (serverIdentity != null) ? serverIdentity : serverPublicKey;
    byte[] cId = (clientIdentity != null) ? clientIdentity : record.clientPublicKey();

    byte[] mn = (maskingNonce != null) ? maskingNonce : config.randomProvider().randomBytes(OpaqueConfig.Nn);
    CredentialRequest credReq = new CredentialRequest(ke1.credentialRequest().blindedElement());
    CredentialResponse credResponse = OpaqueCredentials.createCredentialResponseWithNonce(
        config, credReq, serverPublicKey, record, credentialIdentifier, oprfSeed, mn);

    OpaqueCipherSuite.AkeKeyPair serverAkeKp = suite.deriveAkeKeyPair(serverAkeKeySeed);
    BigInteger serverAkeSk = serverAkeKp.privateKey();
    byte[] serverAkePk = serverAkeKp.publicKeyBytes();

    byte[] preamble = buildPreamble(config.context(), cId, ke1, sId, credResponse, serverNonce, serverAkePk);
    byte[] preambleHash = suite.hash(preamble);

    GroupSpec gs = suite.oprfSuite().groupSpec();
    byte[] dh1 = gs.scalarMultiply(serverAkeSk, ke1.clientAkePublicKey());
    byte[] dh2 = gs.scalarMultiply(serverPrivateKey, ke1.clientAkePublicKey());
    byte[] dh3 = gs.scalarMultiply(serverAkeSk, record.clientPublicKey());
    byte[] ikm = ByteUtils.concat(dh1, dh2, dh3);
    Arrays.fill(dh1, (byte) 0);
    Arrays.fill(dh2, (byte) 0);
    Arrays.fill(dh3, (byte) 0);

    DerivedKeys keys = deriveKeys(config, ikm, preamble);
    Arrays.fill(ikm, (byte) 0);
    byte[] serverMac = suite.hmac(keys.km2(), preambleHash);
    byte[] expectedClientMac = suite.hmac(keys.km3(),
        suite.hash(ByteUtils.concat(preamble, serverMac)));
    Arrays.fill(keys.km2(), (byte) 0);
    Arrays.fill(keys.km3(), (byte) 0);

    ServerAuthState authState = new ServerAuthState(expectedClientMac, keys.sessionKey());
    KE2 ke2 = new KE2(credResponse, serverNonce, serverAkePk, serverMac);
    return new ServerKE2Result(authState, ke2);
  }

  /**
   * Client GenerateKE3.
   *
   * @param state          client auth state from GenerateKE1
   * @param clientIdentity client identity (null = use clientPublicKey from record)
   * @param serverIdentity server identity (null = use serverPublicKey from credential response)
   * @param ke2            server's KE2
   * @param context        application context
   * @param config         OPAQUE configuration
   * @return AuthResult
   */
  public static AuthResult generateKE3(ClientAuthState state, byte[] clientIdentity,
                                       byte[] serverIdentity, KE2 ke2, byte[] context,
                                       OpaqueConfig config) {
    OpaqueCipherSuite suite = config.cipherSuite();

    RecoverResult recovered = OpaqueCredentials.recoverCredentials(
        state.password(), state.blind(), ke2.credentialResponse(),
        serverIdentity, clientIdentity, config);

    byte[] cId = (clientIdentity != null) ? clientIdentity : recovered.clientPublicKey();
    byte[] sId = (serverIdentity != null) ? serverIdentity
        : recovered.cleartextCredentials().serverPublicKey();

    byte[] preamble = buildPreamble(context, cId, state.ke1(), sId,
        ke2.credentialResponse(), ke2.serverNonce(), ke2.serverAkePublicKey());
    byte[] preambleHash = suite.hash(preamble);

    GroupSpec gs = suite.oprfSuite().groupSpec();
    BigInteger clientSk = recovered.clientPrivateKey();

    byte[] dh1 = gs.scalarMultiply(state.clientAkePrivateKey(), ke2.serverAkePublicKey());
    byte[] dh2 = gs.scalarMultiply(state.clientAkePrivateKey(), recovered.cleartextCredentials().serverPublicKey());
    byte[] dh3 = gs.scalarMultiply(clientSk, ke2.serverAkePublicKey());
    byte[] ikm = ByteUtils.concat(dh1, dh2, dh3);
    Arrays.fill(dh1, (byte) 0);
    Arrays.fill(dh2, (byte) 0);
    Arrays.fill(dh3, (byte) 0);

    DerivedKeys keys = deriveKeys(config, ikm, preamble);
    Arrays.fill(ikm, (byte) 0);

    byte[] expectedServerMac = suite.hmac(keys.km2(), preambleHash);
    Arrays.fill(keys.km2(), (byte) 0);
    // Security: constant-time comparison prevents timing side-channel attacks on MAC verification
    if (!MessageDigest.isEqual(expectedServerMac, ke2.serverMac())) {
      Arrays.fill(expectedServerMac, (byte) 0);
      Arrays.fill(keys.km3(), (byte) 0);
      Arrays.fill(keys.sessionKey(), (byte) 0);
      throw new SecurityException("Authentication failed");
    }
    Arrays.fill(expectedServerMac, (byte) 0);

    byte[] clientMac = suite.hmac(keys.km3(),
        suite.hash(ByteUtils.concat(preamble, ke2.serverMac())));
    Arrays.fill(keys.km3(), (byte) 0);

    return new AuthResult(new KE3(clientMac), keys.sessionKey(), recovered.exportKey());
  }

  private static byte[] encodeVector(byte[] data) {
    return ByteUtils.concat(ByteUtils.I2OSP(data.length, 2), data);
  }

  private static byte[] serializeCredentialResponse(CredentialResponse cr) {
    return ByteUtils.concat(cr.evaluatedElement(), cr.maskingNonce(), cr.maskedResponse());
  }

  private record DerivedKeys(byte[] km2, byte[] km3, byte[] sessionKey) {
  }
}

File: hofmann-rust/src/opaque/internal/opaque_ake.rs
use crate::common::{concat, ct_eq, i2osp};
use crate::opaque::config::OpaqueConfig;
use crate::opaque::config::NN;
use crate::opaque::internal::opaque_credentials;
use crate::opaque::model::*;
use zeroize::Zeroize;

/// Keys derived from the 3DH input key material and preamble.
struct DerivedKeys {
    km2: Vec<u8>,
    km3: Vec<u8>,
    session_key: Vec<u8>,
}

impl Drop for DerivedKeys {
    fn drop(&mut self) {
        self.km2.zeroize();
        self.km3.zeroize();
        self.session_key.zeroize();
    }
}

fn encode_vector(data: &[u8]) -> Vec<u8> {
    concat(&[&i2osp(data.len() as u32, 2), data])
}

fn serialize_credential_response(cr: &CredentialResponse) -> Vec<u8> {
    concat(&[
        &cr.evaluated_element,
        &cr.masking_nonce,
        &cr.masked_response,
    ])
}

/// Constructs the preamble for key derivation.
pub fn build_preamble(
    context: &[u8],
    client_identity: &[u8],
    ke1: &KE1,
    server_identity: &[u8],
    credential_response: &CredentialResponse,
    server_nonce: &[u8],
    server_ake_public_key: &[u8],
) -> Vec<u8> {
    concat(&[
        b"OPAQUEv1-",
        &encode_vector(context),
        &encode_vector(client_identity),
        &ke1.serialize(),
        &encode_vector(server_identity),
        &serialize_credential_response(credential_response),
        server_nonce,
        server_ake_public_key,
    ])
}

/// Derives session keys from the triple-DH input key material and preamble.
///
/// Produces `km2` (server MAC key), `km3` (client MAC key), and the
/// shared `session_key` via HKDF-Extract + HKDF-Expand-Label.
fn derive_keys(config: &OpaqueConfig, ikm: &[u8], preamble: &[u8]) -> DerivedKeys {
    let suite = config.cipher_suite();
    let prk = suite.hkdf_extract(&[], ikm);
    let preamble_hash = suite.hash(preamble);

    let handshake_secret =
        suite.hkdf_expand_label(&prk, b"HandshakeSecret", &preamble_hash, config.nx());
    let session_key = suite.hkdf_expand_label(&prk, b"SessionKey", &preamble_hash, config.nx());

    let km2 = suite.hkdf_expand_label(&handshake_secret, b"ServerMAC", &[], config.nm());
    let km3 = suite.hkdf_expand_label(&handshake_secret, b"ClientMAC", &[], config.nm());

    DerivedKeys {
        km2,
        km3,
        session_key,
    }
}

/// Server GenerateKE2 with deterministic nonces and seeds.
#[allow(clippy::too_many_arguments)]
pub fn generate_ke2_deterministic(
    config: &OpaqueConfig,
    server_identity: Option<&[u8]>,
    server_private_key: &[u8],
    server_public_key: &[u8],
    record: &RegistrationRecord,
    credential_identifier: &[u8],
    oprf_seed: &[u8],
    ke1: &KE1,
    client_identity: Option<&[u8]>,
    masking_nonce: &[u8],
    server_ake_key_seed: &[u8],
    server_nonce: &[u8],
) -> Result<ServerKE2Result, &'static str> {
    let suite = config.cipher_suite();

    let s_id = server_identity.unwrap_or(server_public_key);
    let c_id = client_identity.unwrap_or(&record.client_public_key);

    let cred_req = CredentialRequest {
        blinded_element: ke1.credential_request.blinded_element.clone(),
    };
    let cred_response = opaque_credentials::create_credential_response_with_nonce(
        config,
        &cred_req,
        server_public_key,
        record,
        credential_identifier,
        oprf_seed,
        masking_nonce,
    )?;

    let server_ake_kp = suite.derive_ake_key_pair(server_ake_key_seed);
    let server_ake_sk = server_ake_kp.private_key;
    let server_ake_pk = server_ake_kp.public_key;

    let preamble = build_preamble(
        config.context(),
        c_id,
        ke1,
        s_id,
        &cred_response,
        server_nonce,
        &server_ake_pk,
    );

    // dh1/dh2 multiply the client-supplied AKE public key, so a malformed or
    // identity key returns Err instead of panicking the server thread.
    let gs = suite.oprf_suite().group_spec();
    let dh1 = gs.scalar_multiply(&server_ake_sk, &ke1.client_ake_public_key)?;
    let dh2 = gs.scalar_multiply(server_private_key, &ke1.client_ake_public_key)?;
    let dh3 = gs.scalar_multiply(&server_ake_sk, &record.client_public_key)?;
    let ikm = concat(&[&dh1, &dh2, &dh3]);

    let mut keys = derive_keys(config, &ikm, &preamble);
    let preamble_hash = suite.hash(&preamble);
    let server_mac = suite.hmac(&keys.km2, &preamble_hash);
    let expected_client_mac =
        suite.hmac(&keys.km3, &suite.hash(&concat(&[&preamble, &server_mac])));

    let auth_state = ServerAuthState {
        expected_client_mac,
        session_key: std::mem::take(&mut keys.session_key),
    };
    let ke2 = KE2 {
        credential_response: cred_response,
        server_nonce: server_nonce.to_vec(),
        server_ake_public_key: server_ake_pk,
        server_mac,
    };

    Ok(ServerKE2Result {
        server_auth_state: auth_state,
        ke2,
    })
}

/// Server GenerateKE2 with random nonces.
#[allow(clippy::too_many_arguments)]
pub fn generate_ke2(
    config: &OpaqueConfig,
    server_identity: Option<&[u8]>,
    server_private_key: &[u8],
    server_public_key: &[u8],
    record: &RegistrationRecord,
    credential_identifier: &[u8],
    oprf_seed: &[u8],
    ke1: &KE1,
    client_identity: Option<&[u8]>,
    rng: &mut dyn rand_core::CryptoRng,
) -> Result<ServerKE2Result, &'static str> {
    let mut masking_nonce = vec![0u8; NN];
    rng.fill_bytes(&mut masking_nonce);
    let mut server_ake_key_seed = vec![0u8; NN];
    rng.fill_bytes(&mut server_ake_key_seed);
    let mut server_nonce = vec![0u8; NN];
    rng.fill_bytes(&mut server_nonce);

    generate_ke2_deterministic(
        config,
        server_identity,
        server_private_key,
        server_public_key,
        record,
        credential_identifier,
        oprf_seed,
        ke1,
        client_identity,
        &masking_nonce,
        &server_ake_key_seed,
        &server_nonce,
    )
}

/// Client GenerateKE3: recovers credentials, verifies server MAC, computes client MAC.
///
/// Returns the client's KE3 message, session key, and export key on success.
/// Returns `Err` if server MAC verification fails (wrong password or tampered KE2).
pub fn generate_ke3(
    state: &ClientAuthState,
    client_identity: Option<&[u8]>,
    server_identity: Option<&[u8]>,
    ke2: &KE2,
    context: &[u8],
    config: &OpaqueConfig,
) -> Result<AuthResult, &'static str> {
    let suite = config.cipher_suite();

    let recovered = opaque_credentials::recover_credentials(
        &state.password,
        &state.blind,
        &ke2.credential_response,
        server_identity,
        client_identity,
        config,
    )?;

    let c_id = client_identity.unwrap_or(&recovered.client_public_key);
    let s_id = server_identity.unwrap_or(&recovered.cleartext_credentials.server_public_key);

    let preamble = build_preamble(
        context,
        c_id,
        &state.ke1,
        s_id,
        &ke2.credential_response,
        &ke2.server_nonce,
        &ke2.server_ake_public_key,
    );

    // These multiply server-supplied public keys (the ephemeral key from KE2 and
    // the static key recovered from the masked credential response), so a
    // malformed or identity element returns Err instead of panicking the client.
    let gs = suite.oprf_suite().group_spec();
    let dh1 = gs.scalar_multiply(&state.client_ake_private_key, &ke2.server_ake_public_key)?;
    let dh2 = gs.scalar_multiply(
        &state.client_ake_private_key,
        &recovered.cleartext_credentials.server_public_key,
    )?;
    let dh3 = gs.scalar_multiply(&recovered.client_private_key, &ke2.server_ake_public_key)?;
    let ikm = concat(&[&dh1, &dh2, &dh3]);

    let mut keys = derive_keys(config, &ikm, &preamble);
    let preamble_hash = suite.hash(&preamble);
    let expected_server_mac = suite.hmac(&keys.km2, &preamble_hash);

    if !ct_eq(&expected_server_mac, &ke2.server_mac) {
        return Err("Authentication failed");
    }

    let client_mac = suite.hmac(
        &keys.km3,
        &suite.hash(&concat(&[&preamble, &ke2.server_mac])),
    );

    Ok(AuthResult {
        ke3: KE3 { client_mac },
        session_key: std::mem::take(&mut keys.session_key),
        export_key: recovered.export_key,
    })
}
